addon/authenticators/oauth2-password-grant.js

```diff
@@ -391,7 +409,12 @@ export default BaseAuthenticator.extend({
               response.responseJSON = json;
               reject(response);
             } else {
-              resolve(json);
+              if (resolveWithResponse) {
+                response.responseJSON = json;
```

Comment:
I don't think we should be doing this here. The `response` has a `json` method that should be used instead anyway.

Reply:
So it would be `resolve(response)` and the consumer of the response could use `response.json()` to get JSON instead of using `responseJSON`?

Reply:
Yes, exactly - we shouldn't modify the original response object, that was a bad idea when I first introduced it above.

Reply:
Hmmm. It doesn't appear that you can `resolve(response)` from within the promise returned from `response.text()`, as calling `response.json()` throws errors like `Uncaught TypeError: Already read`. This is caused from calling more than one Body method on a response.

I tried
```
fetch(url, options).then((response) => {
  if (resolveWithResponse) {
    resolve(response);
  } else {
    response.text().then((text) => {
      ...
    }
  }
}
```
but that would cause issues with everywhere that uses `makeRequest`.

I'm wondering if sticking with `response.responseJSON = json;` for now is the way to go as it seems to me that doing it any other way would be pretty invasive.

Reply:
Hm, maybe it's possible to rewind the already read properties? I **think** I remember we did that in some project where we needed to read multiple times… Or maybe we can clone the response before reading and then resolve with the original one?

The problem with adding a `responseJSON` property is that changes the API from _"returns `fetch`'s response"_ to _"returns `fetch`'s response with an additional `responseJSON`"_ property. That might seem trivial but it's something we need to maintain over time as well as something that could break, e.g. if `fetch` ever introduced its own `responseJSON` property that was somehow different from ours though.

Reply:
I'm not sure how to "rewind" the read properties, but we can `clone` the response. The only real issue with doing that is all the places that use `makeRequest` have to be able to handle JSON or Response data, which makes this PR more invasive than I intended. I'll see how `clone` pans out.

I'm not sure the direction that this project is going, but I could see a case for maintaining `responseJSON` until the next major release. At that point the API could be changed where the response is only a `Response` object.

Reply:
This change will introduce what's going to be supported in the next major release so we should make sure it's what we want to maintain for the long term. I'd clone the `response` - that seems to be a pretty reasonable approach.